pandas/indexes/base.py

```diff
@@ -2630,6 +2630,20 @@ def _join_multi(self, other, how, return_indexers=True):
         raise NotImplementedError("merging with both multi-indexes is not "
                                   "implemented")
 
+    def _array_putmask(self, a, mask, values):
+        """
+        Must be overridden by Indexes which have non-ndarray .values (e.g.
+        CategoricalIndex).
+        """
+        np.putmask(a, mask, values)
+
+    def _join_values_take_kwargs(self, other):
+        """
+        Must be overridden by Indexes which have non standard
+        self.values.take() keyword args (e.g. CategoricalIndex).
+        """
```

Comment:
... as already said, if you want me I override the entire ``_join_non_unique`` I will, I don't see many other alternatives currently.